core/src/main/java/com/sequenceiq/cloudbreak/converter/v4/environment/network/EnvironmentBaseNetworkConverter.java

```diff
@@ -32,34 +34,37 @@ public Network convertToLegacyNetwork(EnvironmentNetworkResponse source, String
         result.setOutboundInternetTraffic(source.getOutboundInternetTraffic());
         result.setNetworkCidrs(source.getNetworkCidrs());
         Map<String, Object> attributes = new HashMap<>();
-        Optional<CloudSubnet> cloudSubnet;
-        if (StringUtils.isNotBlank(source.getPreferedSubnetId())) {
-            LOGGER.debug("Choosing subnet by prefered subnet Id {}", source.getPreferedSubnetId());
-            CloudSubnet cloudSubnetById = source.getSubnetMetas().get(source.getPreferedSubnetId());
-            if (cloudSubnetById == null) {
-                cloudSubnetById = source.getSubnetMetas().values()
-                        .stream()
-                        .filter(e -> e.getId().equals(source.getPreferedSubnetId()))
-                        .findFirst().orElse(null);
-            }
-            cloudSubnet = Optional.of(cloudSubnetById);
-        } else if (StringUtils.isNotEmpty(availabilityZone)) {
-            LOGGER.debug("Choosing subnet by availability zone {}", availabilityZone);
-            cloudSubnet = source.getSubnetMetas().values().stream()
-                    .filter(s -> StringUtils.isNotEmpty(s.getAvailabilityZone()) &&
-                            s.getAvailabilityZone().equals(availabilityZone))
-                    .findFirst();
-        } else {
-            LOGGER.debug("Fallback to choose random subnet");
-            cloudSubnet = source.getSubnetMetas().values().stream().findFirst();
-        }
+        Optional<CloudSubnet> cloudSubnet = chooseSubnet(source.getPreferedSubnetId(), source.getSubnetMetas(), availabilityZone, true);
         if (cloudSubnet.isEmpty()) {
             throw new BadRequestException("No subnet for the given availability zone: " + availabilityZone);
         }
         LOGGER.debug("Chosen subnet: {}", cloudSubnet.get());
+
+        if (source.getPublicEndpointAccessGateway() == PublicEndpointAccessGateway.ENABLED) {
+            String selectedAZ = cloudSubnet.get().getAvailabilityZone();
+            Map<String, CloudSubnet> subnetsToParse;
+            if (source.getGatewayEndpointSubnetMetas() == null || source.getGatewayEndpointSubnetMetas().isEmpty()) {
+                subnetsToParse = source.getSubnetMetas();
+            } else {
+                subnetsToParse = source.getGatewayEndpointSubnetMetas();
+            }
+            Map<String, CloudSubnet> publicSubnetMetas = subnetsToParse.entrySet().stream()
+                .filter(entry -> !entry.getValue().isPrivateSubnet())
+                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+
+            Optional<CloudSubnet> endpointGatewayCloudSubnet = chooseSubnet(null,
+                publicSubnetMetas, selectedAZ, false);
```

Comment:
Should we still use the `selectedAZ` here? The selectedAZ can be coming from the network request, but the user maybe explicitly specifying the subnet to use for the endpoint gateway anyways if provided. I was thinking that the provided getGatewayEndpointSubnetMetas will always take precedence.

Reply:
I'd argue against the getGatewayEndpointSubnetMetas taking precedence, since their use case is very limited vs. the private subnets which are used more widely. But we do need to include the selected AZ, because the private and public subnets needs to be in the same AZ for the load balancer to work.